Sources/StreamChat/Database/DTOs/MessageDTO.swift

```diff
@@ -550,6 +550,82 @@ extension NSManagedObjectContext: MessageDatabaseSession {
         message.pinnedBy = nil
         message.pinExpires = nil
     }
+
+    /// Adds the reaction for the current user to the message with id `messageId`
+    ///
+    /// Notes:
+    /// - The reaction is added to the database and it updates the message `reactionScores` property
+    /// - This method will throw if there is no current user set
+    /// - If the message is not found, there will be no side effect and the method will return `nil`
+    /// - If a reaction for the same user, type and message exists
+    func addReaction(
+        to messageId: MessageId,
+        type: MessageReactionType,
+        score: Int,
+        extraData: [String: RawJSON]
+    ) throws -> MessageReactionDTO? {
+        guard let currentUserDTO = currentUser else {
+            throw ClientError.CurrentUserDoesNotExist()
+        }
+
+        guard let message = MessageDTO.load(id: messageId, context: self) else {
+            return nil
+        }
+
+        let result = try MessageReactionDTO.loadOrCreate(messageId: messageId, type: type, user: currentUserDTO.user, context: self)
+
+        // make sure we update the reactionScores for the message in a way that works for new or updated reactions
+        let scoreDiff = Int64(score) - result.dto.score
+        let newScore = max(0, message.reactionScores[type.rawValue] ?? Int(result.dto.score) + Int(scoreDiff))
+        message.reactionScores[type.rawValue] = newScore
+
+        result.dto.score = Int64(score)
+        result.dto.extraData = try JSONEncoder.default.encode(extraData)
+
+        if message.reactions.filter({ MessageReactionDTO.createId(dto: $0) == MessageReactionDTO.createId(dto: result.dto) })
+            .isEmpty {
+            message.reactions.insert(result.dto)
```

Comment:
Seems that adding a reaction to `message.reactions` is enough. This `Message ->> Reactions` relation is a set so the attempt to add an already existing item will take no effect and can be done without a prior check.
```suggestion
message.reactions.insert(result.dto)
```